### PR TITLE
Add ComplexParams readable and writable

### PR DESCRIPTION
**Pull request by mhamilton723**
- Add a set of serialization traits that work with arbitrary parameter types
- Fix CNTKModel/ImageFeaturizer so that byteArrays are not stored twice

### CODE REVIEW

File: src/cntk-model/src/main/scala/CNTKModel.scala [hunk 1]
```diff
@@ -130,34 +131,31 @@ object CNTKModel extends DefaultParamsReadable[CNTKModel] {
     model
   }
 
-  override def load(path: String): CNTKModel = super.load(path)
 }
 
-trait CNTKModelParam extends Wrappable {
+@InternalWrapper
```
Comment (mhamilton723) [inline src/cntk-model/src/main/scala/CNTKModel.scala]: Why is this here

File: src/codegen/src/main/scala/PySparkWrapperTest.scala [hunk 2]
```diff
@@ -174,7 +174,6 @@ abstract class PySparkWrapperTest(entryPoint: PipelineStage,
         case "FastVectorAssembler" => "inputCols=\"col1\""
         case "MultiNGram"          => "inputColumns=np.array([ \"col5\", \"col6\" ])"
         case "SelectColumns"       => "cols=[\"col1\"]"
-        case "ImageFeaturizer"     => "modelSaveDir=\"file:///tmp\""
```
Comment (drdarshan) [inline src/codegen/src/main/scala/PySparkWrapperTest.scala]: Please check why you needed to delete this..

File: src/core/serialize/src/main/scala/params/ArrayMapParam.scala [hunk 3]
```diff
@@ -3,13 +3,16 @@
...
 
-import spray.json._
 import org.apache.spark.ml.util.Identifiable
+import spray.json.{DefaultJsonProtocol, _}
+
 import scala.collection.immutable.Map
 
+/**
+  * Created by marhamil on 7/21/17.
+  */
```
Comment (drdarshan) [inline src/core/serialize/src/main/scala/params/ArrayMapParam.scala]: Remove?
